File: frontend/src/lib/paper-trade-math.ts
import type {
  PaperTrade,
  PaperTradePeriod,
  PaperTradeStats,
  PaperTradeStatus,
  PaperTradeTakeProfitTarget,
} from "../types/paper-trade";

export function isPaperTradeOpen(trade: PaperTrade): boolean {
  return trade.status === "open";
}

export function unrealizedPnlUsd(trade: PaperTrade, currentPrice: number): number {
  if (!isPaperTradeOpen(trade) || currentPrice <= 0) return 0;
  const diff =
    trade.side === "long" ? currentPrice - trade.entryPrice : trade.entryPrice - currentPrice;
  return diff * trade.sizeBtc;
}

export function realizedPnlFromExit(
  side: PaperTrade["side"],
  entryPrice: number,
  exitPrice: number,
  sizeBtc: number,
): number {
  const diff = side === "long" ? exitPrice - entryPrice : entryPrice - exitPrice;
  return diff * sizeBtc;
}

export interface PaperTradeResolution {
  exitPrice: number;
  status: Exclude<PaperTradeStatus, "open" | "closed_manual">;
}

/** Check if current price triggers SL or the selected TP for an open position. */
export function resolvePaperTradeExit(
  trade: PaperTrade,
  price: number,
): PaperTradeResolution | null {
  if (!isPaperTradeOpen(trade) || price <= 0) return null;

  const { side, stopLoss, takeProfit1, takeProfit2 } = trade;
  const target = trade.takeProfitTarget ?? "tp1";

  if (side === "long") {
    if (price <= stopLoss) return { exitPrice: stopLoss, status: "closed_sl" };
    if (target === "tp2" && takeProfit2 != null && price >= takeProfit2) {
      return { exitPrice: takeProfit2, status: "closed_tp2" };
    }
    if (target === "tp1" && takeProfit1 != null && price >= takeProfit1) {
      return { exitPrice: takeProfit1, status: "closed_tp1" };
    }
    return null;
  }

  if (price >= stopLoss) return { exitPrice: stopLoss, status: "closed_sl" };
  if (target === "tp2" && takeProfit2 != null && price <= takeProfit2) {
    return { exitPrice: takeProfit2, status: "closed_tp2" };
  }
  if (target === "tp1" && takeProfit1 != null && price <= takeProfit1) {
    return { exitPrice: takeProfit1, status: "closed_tp1" };
  }
  return null;
}

export function takeProfitTargetLabel(target: PaperTradeTakeProfitTarget): string {
  return target === "tp1" ? "TP1" : "TP2";
}

export function filterPaperTradesByPeriod(
  trades: PaperTrade[],
  period: PaperTradePeriod,
  now = new Date(),
): PaperTrade[] {
  if (period === "all") return trades;

  const start = new Date(now);
  if (period === "today") {
    start.setHours(0, 0, 0, 0);
  } else if (period === "week") {
    start.setDate(start.getDate() - 7);
  } else {
    start.setDate(start.getDate() - 30);
  }

  return trades.filter((t) => {
    const anchor = t.closedAt ?? t.openedAt;
    return anchor != null && anchor >= start;
  });
}

export function summarizePaperTrades(trades: PaperTrade[]): PaperTradeStats {
  const closed = trades.filter((t) => !isPaperTradeOpen(t));
  const openCount = trades.length - closed.length;

  let winCount = 0;
  let lossCount = 0;
  let breakevenCount = 0;
  let totalRealizedPnlUsd = 0;
  const wins: number[] = [];
  const losses: number[] = [];

  for (const t of closed) {
    const pnl = t.realizedPnlUsd ?? 0;
    totalRealizedPnlUsd += pnl;
    if (pnl > 0.01) {
      winCount += 1;
      wins.push(pnl);
    } else if (pnl < -0.01) {
      lossCount += 1;
      losses.push(pnl);
    } else {
      breakevenCount += 1;
    }
  }

  const decided = winCount + lossCount;
  const winRatePct = decided > 0 ? Math.round((winCount / decided) * 1000) / 10 : null;

  return {
    openCount,
    closedCount: closed.length,
    winCount,
    lossCount,
    breakevenCount,
    winRatePct,
    totalRealizedPnlUsd: Math.round(totalRealizedPnlUsd * 100) / 100,
    avgWinUsd: wins.length ? Math.round((wins.reduce((a, b) => a + b, 0) / wins.length) * 100) / 100 : null,
    avgLossUsd: losses.length
      ? Math.round((losses.reduce((a, b) => a + b, 0) / losses.length) * 100) / 100
      : null,
  };
}

export function statusLabelJa(status: PaperTradeStatus): string {
  switch (status) {
    case "open":
      return "保有中";
    case "closed_tp1":
      return "TP1決済";
    case "closed_tp2":
      return "TP2決済";
    case "closed_sl":
      return "SL決済";
    case "closed_manual":
      return "手動決済";
    default:
      return status;
  }
}

File: backend/app/services/paper_trade_math.py
from __future__ import annotations

from dataclasses import dataclass
from typing import Literal

PaperTradeSide = Literal["long", "short"]
PaperTradeStatus = Literal["open", "closed_tp1", "closed_tp2", "closed_sl", "closed_manual"]
AutoCloseStatus = Literal["closed_tp1", "closed_tp2", "closed_sl"]
TakeProfitTarget = Literal["tp1", "tp2"]


@dataclass(frozen=True)
class PaperTradeState:
    side: PaperTradeSide
    status: PaperTradeStatus
    entry_price: float
    size_btc: float
    stop_loss: float
    take_profit1: float | None
    take_profit2: float | None
    take_profit_target: TakeProfitTarget = "tp1"


@dataclass(frozen=True)
class PaperTradeResolution:
    exit_price: float
    status: AutoCloseStatus


def is_paper_trade_open(trade: PaperTradeState) -> bool:
    return trade.status == "open"


def realized_pnl_usd(
    side: PaperTradeSide,
    entry_price: float,
    exit_price: float,
    size_btc: float,
) -> float:
    diff = exit_price - entry_price if side == "long" else entry_price - exit_price
    return diff * size_btc


def resolve_paper_trade_exit(trade: PaperTradeState, price: float) -> PaperTradeResolution | None:
    if not is_paper_trade_open(trade) or price <= 0:
        return None

    side = trade.side
    stop_loss = trade.stop_loss
    take_profit1 = trade.take_profit1
    take_profit2 = trade.take_profit2
    target = trade.take_profit_target

    if side == "long":
        if price <= stop_loss:
            return PaperTradeResolution(exit_price=stop_loss, status="closed_sl")
        if target == "tp2" and take_profit2 is not None and price >= take_profit2:
            return PaperTradeResolution(exit_price=take_profit2, status="closed_tp2")
        if target == "tp1" and take_profit1 is not None and price >= take_profit1:
            return PaperTradeResolution(exit_price=take_profit1, status="closed_tp1")
        return None

    if price >= stop_loss:
        return PaperTradeResolution(exit_price=stop_loss, status="closed_sl")
    if target == "tp2" and take_profit2 is not None and price <= take_profit2:
        return PaperTradeResolution(exit_price=take_profit2, status="closed_tp2")
    if target == "tp1" and take_profit1 is not None and price <= take_profit1:
        return PaperTradeResolution(exit_price=take_profit1, status="closed_tp1")
    return None


def status_label_ja(status: AutoCloseStatus) -> str:
    return {
        "closed_tp1": "TP1決済",
        "closed_tp2": "TP2決済",
        "closed_sl": "SL決済",
    }[status]
